[ADD] Module to merge draft or confirmed sales orders

StefanRijnhart opened the pull request.

StefanRijnhart committed:
[ADD] Module to merge draft or confirmed sales orders

pedrobaeza commented on sale_order_merge/README.rst:
```diff
@@ -0,0 +1,70 @@
+This module allows the sale employee to merge draft or confirmed orders
+from the same customer.
+
+When orders are merged, draft invoices and unprocessed outgoing pickings
```
Why merging pickings? You can leave them unmerged as Odoo standard allows it.

StefanRijnhart replied:
Why else merge the orders? This was the main use case of my customer.

pedrobaeza replied:
Yeah, you're right. You can't merge pickings without merging the "upstream" sales orders, or the links will not be totally correct (although the main control is through sales order lines, not through sales orders, but the main information/reports won't be correct).

JordiBForgeFlow commented on sale_order_merge/models/sale_order_merge.py:
```diff
@@ -0,0 +1,190 @@
+        for picking in self.sale_order.picking_ids:
+            if (picking.state not in ('done', 'cancel') and
+                    picking.location_dest_id.usage == 'customer'):
+                key = (picking.picking_type_id, picking.location_id,
```
Looking towards extensibility, please build the key in a separate method. We work, for instance, on another model that various pickings for a sales order, based on the requested date maintained at sales order line level. In this case, when we merge orders we'd want to respect the scheduled date.

StefanRijnhart replied:
Done

JordiBForgeFlow commented on sale_order_merge/models/sale_order.py:
```diff
@@ -0,0 +1,74 @@
+            ('partner_shipping_id', '=', sale.partner_shipping_id.id),
+            ('warehouse_id', '=', sale.warehouse_id.id),
+            ('state', 'in', self.MERGE_STATES),
+        ] + policy_clause
```
Looking towards extensibility, please build the domain in a separate method. For example, in sale_operating_unit (https://github.com/OCA/operating-unit/pull/22) you would not want to merge sales orders that belong to separate operating units. Also, why is the company_id not included?

StefanRijnhart replied:
Done

JordiBForgeFlow commented on sale_order_merge/models/sale_order_merge.py:
```diff
@@ -0,0 +1,190 @@
+                target = pickings[0]
+                pickings -= target
+                pickings.move_lines.write({'picking_id': target.id})
+                pickings.unlink()
```
In the context of real time inventory valuation this will cause that the journal entries created from the partial transfers of pickings will refer to a deleted picking (https://github.com/odoo/odoo/blob/8.0/addons/stock_account/stock_account.py#L278)
Perhaps it would be a good idea to add here a hook (at least) that would allow to update these journal entries.

StefanRijnhart replied:
Only draft pickings are merged and deleted, so this is not an issue right?

JordiBForgeFlow commented:
I create a SO 1, I confirm and creates a DO1.
I create a SO2 and while in draft, I merge SO2 into SO1.
A DO2 is created, but not merged with DO1.

![image](https://cloud.githubusercontent.com/assets/7683926/18941315/6731918e-8610-11e6-99b3-d4321d148a15.png)

StefanRijnhart commented:
@jbeficent cannot seem to reply directly to https://github.com/OCA/sale-workflow/pull/341#issuecomment-250369217, but I cannot reproduce this. What are the order settings of SO1 and SO2?

StefanRijnhart commented:
@jbeficent found them on the runbot

StefanRijnhart commented:
@jbeficent You got multiple DO's for the merged order because the runbot instance has the module 'sale procurement group per line' installed. I uninstalled it and it works now.

StefanRijnhart committed:
[ADD] Extensibility hooks
[FIX] Add company to merge domain
[FIX] Fix syntax error in case multiple pickings are merged

JordiBForgeFlow commented on sale_order_merge/models/sale_order_merge.py:
```diff
@@ -0,0 +1,198 @@
+                pickings -= target
+                pickings.mapped('move_lines').write({'picking_id': target.id})
+                pickings.unlink()
+        return True
```
The final picking should have as Source Document the final sales order, not the original.

![image](https://cloud.githubusercontent.com/assets/7683926/18946325/a980bee8-862e-11e6-99c1-4b75aebfe95e.png)

StefanRijnhart replied:
Done

JordiBForgeFlow commented on sale_order_merge/README.rst:
```diff
@@ -0,0 +1,70 @@
+Remove the orders that you do not want to merge, and click on the *Merge*
+button in the footer of the pop-up window. The main window will then refresh
+to the updated main sale order.
+
```
I suggest to add the following:

The mergeability criteria is defined as follows:
- Same customer, shipping address, warehouse and company.
- Orders in status 'Draft Quotation', 'Quotation Sent', 'Waiting Shedule', 'Sale Order', 'Sale to Invoice'.
- Once the order has already been confirmed, only draft orders or confirmed orders with the same invoice policy can be merged.

The criteria can easily be extended using method `_get_merge_domain` of model `sale.order`

StefanRijnhart replied:
Done. Thanks for the review BTW.

StefanRijnhart committed:
[FIX] Update the origin of the remaining pickings
[UPD] README, courtesy of Jordi Ballester

StefanRijnhart committed:
[FIX] Syntax error

StefanRijnhart commented:
Hi @jbeficient, would you like to revise your review based on my changes?

JordiBForgeFlow commented:
@StefanRijnhart This has all the required approvals. Can I merge now?

StefanRijnhart commented:
As it has the two require approvals, yes!

pedrobaeza commented:
Remember to squash on merge as this case there is no significant different commits

StefanRijnhart commented:
Thanks ;-)
